### ISSUE TITLE
How to update parameter manually with gluon in training loop

### ISSUE THREAD

**Issue by shuokay** [apache/incubator-mxnet#14735]
I want to update parameters in the training loop,
```python
for batch in enumerate(train_data):
    forward
    backward
    trainer.step(batch_size)
    mx.nd.waitall()
    params = net.collect_params()
    my_param = params["my_param"]._reduce()  # my_param is update in cpp side.
    do something with my_param
    params["my_param"].set_data(my_param*0)
    mx.nd.waitall()
```
If run in one gpu, the code works well,
If run in muti gpu and in NativeEngine mode, the code works well,
If move ` trainer.step(batch_size)` after `params["my_param"].set_data(my_param*0)`, the code works well,
If run in multi gpu and default ENGINE `my_param` get the incorrect result.
I think maybe I should sync the data after `set_data`, but I did not find the right way.

**Comment by seujung** [comment]
Try to add this line before update the paramter
```
with autograd.pause():
    params["my_param"].set_data(my_param*0)
```

**Comment by frankfliu** [comment]
@mxnet-label-bot add [question, gluon]

**Comment by shuokay** [comment]
@seujung It did not work

**Comment by shuokay** [comment]
@seujung It did not work

**Comment by shuokay** [comment]
If I set `kvstore='local` and `update_on_kvstore=False`, everything works well. I think this problem is caused by kvstore. Is there someone can give any advices?

**Comment by lanking520** [comment]
@roywei Do you have any context?

**Comment by eric-haibin-lin** [comment]
can you set update_on_kvstore=False when creating Trainer? I don't think you need to change the type of the kvstore to local

**Comment by shuokay** [comment]
@eric-haibin-lin I think distribute training need `update_on_kvstore=True`?

**Comment by eric-haibin-lin** [comment]
Both should work after this PR: https://github.com/apache/incubator-mxnet/commit/63ed258063137421e6d4def30435014ab57fb468
